### ISSUE TITLE
Support FlySky i-Bus telemetry

### ISSUE THREAD

**Issue by smiling-Jack** [cleanflight/cleanflight#1125]
Hello
is something planned or in progress for FlySky i-Bus telemetry ?

**Comment by SteveAmor** [comment]
#983 

Please can you close this issue if the above PR answers your question.

**Comment by smiling-Jack** [comment]
i think ist not the same issue or ?

I mean send telemetry data back to the radio. 
e.g. altitude, ....

**Comment by Fheilmann** [comment]
I'd be willing to have a more detailed look at this, however there is absolutely no information on the web on what the sensor part of the i-bus protocol looks like.

**Comment by dixiak** [comment]
I've monitored the connection between an FS-IA6b receiver and an FS-CVT01 add-on voltage sensor. 
The data packets do not look too complicated. However, data transfer is bi-directional on a single wire. Could this be handled on supported unmodified hardware or would we need to add a diode between Rx and Tx? Transfer rate appears to be 115200 so software serial is not likely to work too well.

**Comment by dixiak** [comment]
Most of the time, here's what the two devices say to each other repeatedly.
Receiver says 
04 A1 (5A FF)
Sensor says 
06 A1 xx xx (58 FF)

The bit in brackets is the packet checksum, calculated the same way a regular iBus i.e. 0xffff minus each individual byte value, stored little-endian.

04 A1 is probably a request for a voltage reading. 
06 A1 probably means "here's a voltage reading". The next two bytes are a signed 16-bit integer representing hundreths of a volt. i.e. F401 would be 5.00V (stored little-endian)

**Comment by dixiak** [comment]
There's also a setup sequence, which I have no idea how to interpret.
Receiver repeatedly says
04 81 (7A FF)
04 81 (7A FF)
approx. every 7ms

When you plug in a sensor it will reply with
04 91 (6A FF)

Receiver:
06 91 03 02 (63 FF)

Sensor:
04 82 (79 FF)

Receiver:
04 A1 (5A FF)
Sensor: 
06 A1 xx xx (58 FF)
At this point we have fallen into the pattern described in the previous post, repeated approximately every 6ms.

**Comment by Fheilmann** [comment]
The two-way communication makes this difficult indeed. Maybe someone "official" can chime in about whether we can do 1-wire bidirectional communication on a (soft)-serial port?

**Comment by ledvinap** [comment]
FrSky smartport is bidirectional. It works both on softserial (with heavy changes from master and untested lately) and F3 UART (flying that) in my 'local' branch.
Smartport needs inverted mode (only F3 UART supports this), I-bus should be even easier ...

**Comment by Fheilmann** [comment]
So it may be possible to use the FrSky telemetry port on, say, a NAZE32 to accomplish this? That would be awesome!

**Comment by ledvinap** [comment]
@Fheilmann : There is some SW support for bidirectional communication in CF (on UART), using it for Ibus telemetry should be quite easy

**Comment by Fheilmann** [comment]
@ledvinap Sorry for my stupid questions, I am still trying to get an overview of cleanflight and how everything works.

What kind of port is the special "1-wire frsky telemetry" port that you can find on most naze32 boards?

**Comment by ledvinap** [comment]
IMO it's just transistor-inverted TX signal with pull-up ... Works with FrSky telemetry port (single-direction, just inversion is needed)

**Comment by twokaff** [comment]
I second the request of iBus compatible telemetry support.  LiPo voltage without an additional sensor would be nice.

**Comment by hydra** [comment]
flysky RX support now merged into CF 1.12.0 development versions.

**Comment by CraigJPerry** [comment]
Hey Guys,

I've started working on this tonight. I've proven it works on an F3 discovery (checkout the ludicrous voltage report in this photo - i was just counting up in a loop to test it):

![img_20160122_014700](https://cloud.githubusercontent.com/assets/1908742/12500409/cc1ba05c-c0aa-11e5-96af-93badd088cd4.jpg)

Here's a trace of a higher voltage from the logic analyser:

![voltage sensor logic analyser](https://cloud.githubusercontent.com/assets/1908742/12500361/75b63d8a-c0aa-11e5-9543-c9dd107e6b4f.png)

The protocol works like this:

1. The rx queries: 0x04, 0x81, 0x7A, 0xFF
2. I respond: 0x04, 0x81, 0x7A, 0xFF
3. The rx queries: 0x04, 0x91, 0x6A, 0xFF
4. I respond 0x06 0x91 0x00 0x02 0x66 0xFF
5. At this point the rx seems to be looking for more sensors but i've not explored that yet
6. A few more ms passes then the rx queries 0x04, 0xA1, 0x5A, 0xFF
7. I respond with a value and a checksum, probably easiest expressed in the code:

    static void sendFlySkyVoltage(uint16_t voltage) {
      uint8_t first = voltage;
      uint8_t second = voltage >> 8;
      serialWrite(smartPortSerialPort, 0x06);
      serialWrite(smartPortSerialPort, 0xA1);
      serialWrite(smartPortSerialPort, first);
      serialWrite(smartPortSerialPort, second);

      uint16_t checksum = ~(0x06 + 0xA1 + first + second);
      serialWrite(smartPortSerialPort, checksum & 0xFF);
      serialWrite(smartPortSerialPort, checksum >> 8);
}

Now that i've proven it works, i'll put together a proper patch with test coverage. I just wanted to share that some progress has been made. I've been checking this issue for about a month now - since i started flying small quads, hoping someone would submit a fix!

/me -> bedtime, work in 7 hours...

**Comment by CraigJPerry** [comment]
Here's the temperature gauge protocol kind of working. It displays some data, but not what i was expecting! Anyway, seems it's pretty hot in here so time to logoff for sure!

![img_20160122_021546](https://cloud.githubusercontent.com/assets/1908742/12500799/4ca978fe-c0ae-11e5-90d3-65cfb53762eb.jpg)

**Comment by cklam2** [comment]
Over here Dave Borthwick posted some pic code that emulates a sensor for the i6. Maybe can check what he already figured out:
http://www.rcgroups.com/forums/showpost.php?p=33365816&postcount=152

**Comment by CraigJPerry** [comment]
Nice one! He's documented the fields and the bit I was struggling with - the timing parameters.

I'll get some more time on this over the weekend. It's not a complicated protocol, the code to support it is pretty easy but this is my first stm32 project so I'm sure I'll run into some issues! Just learning about the stm32 support for bi directional half duplex serial has been cool. I thought we'd have to connect the tx / RX pins with a resistor then plug that into the ia6b but there's no need, the stm32 supports this natively. Cool little micro, wayyyy more guts than the arduino's I've been playing with til now!

**Comment by CraigJPerry** [comment]
Hi Guys, can some folks please test PR #1723

Once you've flashed the hex and restored your settings, go to the config tab in the configurator and enable telemetry.

Set the function bitmask to 256 for your chosen serial port. Baud rate is ignored because ibus always runs at 115200. E.g. i run on serial port 1 so i enter this at the cli:

`serial 1 256 115200 57600 115200 115200`

Lastly, to connect it, just use the tx pin of your uart to connect to the ibus sens port. The rx remains unconnected.

Please feedback values you get for "Tem3".

Cheers

**Comment by cklam2** [comment]
Amazing, very excited. 
Trying to make it work but I'm stuck at the last step. Flashed your hex file and in Cleanflight config (ver 1.2.0) trying to set the serial but somehow it doesn't save it. 

```
# serial
serial 0 1 115200 57600 0 115200
serial 1 64 115200 57600 0 115200

# serial 0 257 115200 57600 115200 115200

# save
Saving
Rebooting
```
When checking again it shows this
```
# serial
serial 0 1 115200 57600 0 115200
serial 1 0 115200 57600 0 115200
```
And also lost my serial on UART2. Any tip on this?

**Comment by CraigJPerry** [comment]
Aha IBUS sits and blabs away all the time so if you port share it with MSP like that i think it will cause problems so I didn't enable this sharing for that reason. However i don't know it's a problem for sure i just guessed so we should test it to see, it could be fine to share these. MSP may be robust enough to handle the collisions.

I'll upload a new hex tomorrow night when I'm next on a computer.

Thanks for trying so far! Appreciate it.

**Comment by CraigJPerry** [comment]
If you want to build a new hex yourself before then I also updated the dev docs so that might help some folks too. Also if you have something other than a base. Although, there are much easier ways to build a hex than installing eclipse but this might interest some all the same, help get more contributors to the project!

https://github.com/cleanflight/cleanflight/blob/72e1fab1749ec2a7a14791a27b8fd0f2cb0b6fe9/docs/development/Building%20in%20Eclipse.md

**Comment by Fheilmann** [comment]
I would recommend you get in touch with the people in this rcgroups thread:

http://www.rcgroups.com/forums/showthread.php?t=2486545

They have cracked the Firmware of the FlySky i6 transmitter wide open and I am pretty sure they can help you get meaningful telemetry data onscreen of this TX (they can rename the temperature sensor to something else for example)

**Comment by cklam2** [comment]
I tried that too, built a custom hex file but I see the same thing happening. I actually don't do anything with MSP but when using 256 instead I see the same result. Also when giving up the iBUS rx on UART2 (just want to see telemetry working) and running "serial 1 256 ..." it also resets after I save it. So not sure what I'm doing wrong. Will try again tonight.

**Comment by CraigJPerry** [comment]
Hey @cklam2 

Yeah the serial 0 port being reconfigured to default settings was expected with the last hex, i'd deliberately not marked ibus as shareable. Here's a hex with it shareable:

[cleanflight_NAZE.zip](https://github.com/cleanflight/cleanflight/files/105750/cleanflight_NAZE.zip)

Can you test a) that it works but also b) that the configurator works as normal with this in place. I.e. no random disconnects etc. Make sure there's power to the rx when testing this (i.e. battery plugged in).

Cheers!

**Comment by CraigJPerry** [comment]
@Fheilmann yeah i definitely would, even just to say thanks for some of the stuff those guys have made because it's been awesome for me. However, rcg won't let me login. My original account there (i used to fly rc planes and heli's years ago) got deleted / disabled / disappeared.

My old posts are still on the forums but the account doesn't exist. Super weird. So i created a new account and they blocked me as a spammer! I mailed them for help and i submitted a message via their online form a couple of months ago but never heard anything back.

**Comment by cklam2** [comment]
Hi @CraigJPerry. Sorry, been busy on other things lately. 

I flashed your hex and this is much better, doesn't overwrite the serial 1 when setting the serial 0 for MSP+IBUS telemetry. In the configurator I don't see any random disconnects either.

However, no telemetry info on my TX (flashed with 10ch 1.5 firmware). In the Sensors list I see IntV, Tx.V and Err only. Just to make sure that I followed the right steps:
- [ ] UART1 TX pin (= Naze32 rev5 FrSky [telemetry TX](http://www.processio.com/wp-content/uploads/2015/02/naze32GPS1.png)) connected to FS-iA6B SENS signal pin
- [ ] In configurator, CLI, entered serial 0 257 115200 57600 115200 115200 and then save (verified with serial that setting is indeed saved)
- [ ] In configurator, Configuration, enabled Telemetry + Save and reboot.

Also tried to use UART2 instead by connecting SENS to pin3 and configured serial port 1 but still no luck. In the TX under RX Setup I see a "i-BUS Setup" and it's set to channel CH1, no idea what it means, I left that as-is.

Btw, saw that you successfully posted on RCg with your new account. Hope that Dave or povlhp can provide you valuable information on the iBUS protocol.

**Comment by CraigJPerry** [comment]
Thanks for testing this @cklam2

Can you send me a dump of your CLI please. You've done everything right, it behaves like the pin isn't connected but you're on the correct pin on both the naze and the RX.

**Comment by cklam2** [comment]
Hi @CraigJPerry, Here is the dump of the cli settings ([clidump.txt](https://github.com/cleanflight/cleanflight/files/111291/clidump.txt))

I restored the connection to the UART1 and I connected it as shown in the picture. It's the black wire, bit hard to see. I hope that someone else can try your firmware cause I'm not trusting my iBUS sensor port and configuration anymore. 

![ibus_connection](https://cloud.githubusercontent.com/assets/10433949/12702489/046489b0-c82b-11e5-86d4-1d2e67cf3dd6.jpg)

**Comment by CraigJPerry** [comment]
Your wiring looks good. The only real difference between our setups is you're using IBUS serial RX and I'm stuck on PPM for now. I have a naze32 in the post to me just now so I'll be able to exactly duplicate your setup and test soon!

**Comment by cklam2** [comment]
Well, well @CraigJPerry, I finally got some telemetry data on my RX display :)

Tried everything I could think of but your last post gave me the clue: I only get the i-BUS Telemetry working when configuring PPM instead of i-BUS RX and connecting SENS to pin 3 of UART2. I guess that the UART1 is limited in its use. Or something else need to be modified in the code in order to make it work on this particular port. 

So for now I need to choose between telemetry and 10 channels.

![rx_sensors](https://cloud.githubusercontent.com/assets/10433949/12733090/11aa6310-c939-11e5-8e8c-f6578f31c7ab.jpg)

**Comment by CraigJPerry** [comment]
Nice one @cklam2 ! I suspect this is might be a hardware (naze specific) thing for uart1 since i run both ibus serial 10ch AND ibus telemetry together on an STM32F3DISCOVERY. I'll know more once my naze arrives.

ExtV on sensor 2 shows as 0v for you in that screenshot. I just want to make sure you have vbat monitoring disabled? Or does the battery status show correctly in the cli when you do a "status" command?

Thanks for testing this btw! There's one other person about to test on an fs-i10 then i'll crack on with augmenting the data we can send back.

All the telemetry implementations in cleanflight duplicate their metric gathering logic so I'll re-factor out a common "telemetry source". As well as eliminating that duplication of code it'll make it much easier to write tests if i can inject different scenarios from the tests.

**Comment by cklam2** [comment]
The displayed value is correct cause I did not yet connect the main battery to the vbat connector :) Right now using an external battery monitoring device but I'll switch over to the naze once I have a buzzer.

**Comment by goldjacket** [comment]
From what I have read about the naze32 you can't connect with USB and use UART 1 pin 1 rev 6 board at the same time. You should try this again in ibus mode not ppm to see if this works while not connected to USB at the same time. Use your Lipo only for power.

**Comment by goldjacket** [comment]
Pin 1 is the TX pin uart1 just to clarify

**Comment by AtomoxS3R** [comment]
Hi @CraigJPerry, probe the firmware with the telemetry. And it worked the first time. I have a Skyline32 as flight controller.
No option to upgrade to the latest version of naze32 with telemetry option?
And I wanted to ask a question, do not know much about it, but can be placed in the telemetry option softserial? The only problem is that the port operates at 19200 max.
regards
Sorry for the translation, my language is Spanish and would translate through Google translator.

**Comment by CraigJPerry** [comment]
Hey @AtomoxS3R thanks for testing and i'm glad it worked for you!

>> No option to upgrade to the latest version of naze32 with telemetry option?

I've just changed PC but as soon as i get everything re-installed, i'll make a new build.

>> can be placed in the telemetry option softserial?

No, it can't work on softserial unfortunately. ibus is fixed speed at 115200.

All the best,

Craig

**Comment by hopoffbaby** [comment]
Hi,

First post here. I have just flashed cleanflight onto a SPRacingF3 and set up Serial RX using IBUS. I have the TGY i10 tx and TGY ia10 receiver. I'd like to test your telemetry code too if possible. Could you please point me in the right direction?

Also, do you know if the ia10 is capable of 14 channels? It is supplied with a TGY-AEV01 IBUS module which according to the manual "will allow the user to add 4 more channels if the channels of the receiver is not sufficient". It also suggests that up to 4 modules can be daisy chained to provide an additional 16 channels.

Thanks,
Neil.

**Comment by Kayouu** [comment]
Were you able to get ibus talking to cleanflight with the ia10 rx? If so how did you connect it? @hopoffbaby

**Comment by Arise** [comment]
Do I have to use the second IBUS port on receiver (the IN port)?

**Comment by hopoffbaby** [comment]
@Kayouu I connected it by using the iBUS OUT port and connecting that to the boards input as described in the documentation. It required me to cut the cable on one of my rx sensor modules and solder to the SPRacingF3 harness. I then configured SerialRX and set it to iBUS in the configuration.

@CraigJPerry I assume that for the telemetry I will need to connect to the iBUS IN in the same way?

**Comment by Kayouu** [comment]
@hopoffbaby and by doing this you were able to control your quad using serialrx \ ibus? Do you enable sbus in the RX settings on the i10? And are you running CF firmware or betaflight? Thanks!

**Comment by Arise** [comment]
@Kayouu already explained this to you here: http://www.rcgroups.com/forums/showpost.php?p=34530160&postcount=2120

**Comment by inode-** [comment]
I'm tryning the NAZE32 hex that you provide, I got the telemetry on the UART2, all values are 0 like into the @cklam2 photo, but I will check that later. The main problem for me is that currently when you set the serial port with that mask you remove all settings from the port, so it's not possible to use the same port for the RX. It's possible to setup that in the correct way to have the IBUS send and IBUS receive on the same port? Ex. PIN3 connected to SENS, PIN4 connected to SERVO?

In your opinion will be possible to receive GPS cords via ibus telemetry?

**Comment by Arise** [comment]
Yes, I would +1 the above reply. Need telemetry on some shared port or on same port as IBus.

**Comment by CraigJPerry** [comment]
Ibus telemetry is bi directional. Although it connects with only 1 wire at
the receiver, it's using 2ports at the stm32 - both tx and RX pins are
used. The stm32 internally connects tx and RX pins if you set bidi mode on
the serial port. So although there's only 1 wire connection to the stn32,
internally it's duplexing tx and RX outputs to the one pin. There is no
spare RX pin on the serial port, it is being used by ibus telemetry even if
there is only 1 wire.

**Comment by CraigJPerry** [comment]
@Arise already shareable with msp on the serial port

**Comment by Arise** [comment]
Ok, thanks for some clarifications.

However on naze32 the MSP is set on UART1 and the receiver is connected via the UART2/pin4 (single wire).

Is it possible to move receiver on UART1 pin? So I can share the receiver (rx part and telemetry) with the MSP port?

**Comment by inode-** [comment]
@CraigJPerry do you think that the two ibus are sharable on the same port?

Another thing that I saw, others telemetry code send the data only when the drone is armed, in this way doesn't create problem when the drone is connected to cleanflight. Maybe this can be used as a workaround if it cause problem on the UART1 port.

**Comment by CraigJPerry** [comment]
I don't think both ibus's can share 1 port. You'd need some changes on the
software side but I think that could be done. Electrically though on the
hardware side of things - I'm really not sure how you would connect the
extra transmitting node without risking damage. You couldn't just use a
resistor it'd skew the signal. It'd need to be some transistor based
switching you put in the middle.

**Comment by AtomoxS3R** [comment]
Hello @CraigJPerry, applying some new telemetry to the most current version of cleanflight for Naze. There is no way it is applied standard for future versions of the .hex
And it could be applied telemetry code in the code of 10ch to unify the tx and rx in ibus.
I have no idea of programming, sorry for this last question if it is a little strange.
We depend on you, we do not know the programming. Thank you.

**Comment by Arise** [comment]
@CraigJPerry  I took a look over your commits and I can see:
```
 * Due to the limited telemetry data types possible with ibus, we
 * simply send everything which can be represented. Currently this
 * is voltage and temperature.

...

static const uint8_t SENSOR_ADDRESS_TYPE_LOOKUP[] = {
    IBUS_SENSOR_TYPE_INTERNAL_VOLTAGE,  // Address 0, not usable since it is reserved for internal voltage
    IBUS_SENSOR_TYPE_EXTERNAL_VOLTAGE,  // Address 1, VBAT
    IBUS_SENSOR_TYPE_TEMPERATURE,       // Address 2, Gyro Temp
    IBUS_SENSOR_TYPE_RPM                // Address 3, Throttle command
};

```
What happens if we add more sensors? Could there be hidden sensors that could be processed?

Is it possible to send data like the Altitude or GPS coordinates? Maybe send them unethically disguised as RPM/Voltage/Temperature? Even if the TX will report them as wrong, we could know what their meaning is...

**Comment by hopoffbaby** [comment]
Hi,

Does anyone know how I can compile this to work with SPRacingF3 or point me to some documentation?

Thanks,
Neil.

**Comment by CraigJPerry** [comment]
Hi @Arise, you can freely add more sensors. Addresses 0-15 are supported by the protocol. There are no hidden sensors.

It's possible to send other data types, i asked if the guys doing the FSi6 firmware patches over on RCGroups would mind labelling a few new datatypes like GPS co-ords, height, distance from home etc. or even just a generic "data" type.

I don't have time to chase that up right now, i just occasionally reply on Github notifications. When i get back to rc quads i expect to have more time for this.

Cheers,

**Comment by unitware** [comment]
Hi, I've made some more work based on @CraigJPerry :s pull request. baiscally rebased it on the release v1.13 and added unit tests. Now I'm working on supporting daisychained sensors between the RX and the flightcontroller. However this is a bit if blind flying for me as I do not have the sensors myself. I have made the assumption that every unit in the chain only resends the adresses not used further down the line, which might or might not be valid.

Could anyone help me out and test this code on a Naze32? (other targets can be built on request of course) 

set serial 512 1 115200 57600 115200 115200

If it does not work then it would be very helpful to get a serial dump from the signal line at various points in the daisy chain.

(Do not fly this one, I have not. And make a backup as always):
https://www.dropbox.com/s/d8r04bb36318hi2/cleanflight_NAZE_daisy_chainable.hex?dl=0

**Comment by Arise** [comment]
Almost forgot about iBUS extension, I can give it a try tomorrow. What exactly should I test? You want to daisy chain real sensors or to just see if the TX can see the voltage?

**Comment by unitware** [comment]
Real sensors in daisy chain with the fc at the end.

**Comment by Arise** [comment]
Can you explain how you wired (connected) the sensors and the FC?
Usually for naze32 the receiver connects to the PIN4 but how do you daisy chains the rest of the sensors?

Source code so I can try your builds?

**Comment by unitware** [comment]
Hi, the code is here:
https://github.com/unitware/cleanflight/tree/ibus-telemetry

I do not possess any sensors, hence the request for a test run.
My understanding is that the signal normally going to pin4 on FC should be routed via the pyshical sensors - ie daisy chaining. so the RX in one end, connected to one side of the sensor, then the other side of the sensor to next sensor, and then finally to the FC pin 4.

**Comment by unitware** [comment]
@Arise did you have any luck testing this?

**Comment by unitware** [comment]
I just created an updated pull request https://github.com/cleanflight/cleanflight/issues/2415 with my additions to this one. Thanks for the baseline @CraigJPerry et.al. !

**Comment by unitware** [comment]
Hi, I would really appreciate if some of you would like to make a test flight with pull request https://github.com/cleanflight/cleanflight/pull/2415 and verify that its in working order. 

I have tested it myself but I'm not that experienced in flying and my PIDs are not tuned so I wouldn't see if there was some slight stability change.

**Comment by povlhp** [comment]
So how do we connect things ? The source only says the conect the Tx pin to Rx.
Can we share Rx/tx on the same port ? I am on F3, so I have serial ports to use. But wiring stuff is what I need info on. I think the sensor data requests was not on the iBus data port used for Tx data -> FC.
Is the Tx pin running in bidirectional mode like the source code indicates ?

Can't figure it out, and I do know something about this stuff. I made the iBus2PPM converter project.

**Comment by unitware** [comment]
@povlhp 
https://github.com/cleanflight/cleanflight/blob/master/docs/Telemetry.md#ibus-telemetry

just connect the FC TX pin to the ibus data pin. The rx pin cannot be used for serial when the port is in bidirectional mode.

**Comment by povlhp** [comment]
FC Tx to the iBus sensor data bin ? And another from iBus Servo to the Rx
on another UART ? Or can we do with one data wire ?

**Comment by unitware** [comment]
yes, and no - we cannot share the uart for serial RX and ibus telemetry at this time. (it might be possible with some dev effort and some more hardware)

**Comment by jakubk** [comment]
Hello, 
I am able to enable handling of additional sensors in FS-i6 - currently without any data formatting and names - this list must be dynamic. If someone want to test if it is working please let me know.

**Comment by marcroe** [comment]
Hi @jakubk I would be interested. At the moment I don't have a board that runs CF, only BF and RF. (I destroyed my SPRACINGF3 in a crash a few days ago).
Over at Betaflight, we are also trying to make iBus telemetry work, see: https://github.com/marcroe/betaflight/tree/iBusTelem
Maybe I can still help out somehow.

**Comment by jakubk** [comment]
@marcroe Great - I currently don't have any STM32 based FC - because of that I am asking you guys for help. Basically I am able to edit sensors names and method formatting sensor value as well, so theoretically It is possible to merge 2byte sensor data into 4 bytes format - for example if "gps" sensor needs 4 bytes to express latitude we can transfer this data as sensors with some ID and then 2 indexes must be used. They Key is to change formatting method to get this data. All sensor data is stored in TX as 256 byte array each sensor is express as [ID][Index][val][val]. Defaults sensors are using 24 bytes so there is a lot of empty space. Currently I am adding some dummy names to TX firmware to be able to see if we get different sensors working. Formatting method will be tricky one because we have to know what sensor id is meaning and how to handle it. If you want to flash your TX with my firmware please let me know. It is already published on RC groups but it needs some improvements.

**Comment by AleksanderAzizi** [comment]
@jakubk Would this allow for voltage telemetry from the FC? If so I'm more than ready to test it.

**Comment by goldjacket** [comment]
yes toss me the firmware link and I'll give it a go. also are you using CF
1.2.4 and as for settings do I enable RSSI? or any other settings for it to
show up in OSD. thanks

**Comment by goldjacket** [comment]
what the rc group link can't seem to find the related post.

thanks

**Comment by unitware** [comment]
@AleksanderAzizi Voltage telemetry (as well as temperature and throttle value) is possible with the latest CF release and stock FS-i6 transceiver firmware.

@jakubk I guess gps ccordinates could be very useful, or distance and bearing from starting point. Altitude interesting also. I'd love to test that but first I have another question for you.

 - Do you have insight in the receiver firmware? If so, would it be possible to route the serialRx packet to the telemetry serial port? Or if not can it be confirmed that the telemetry and serial rx is not active at the same time? It looks like it on my logic analyser. What I'm interested in is to combine these serial ports on the FC side, at the moment they need two ports. (I'm working on a prototype firmware to combine these signals in the FC)

**Comment by jakubk** [comment]
@unitware I have disassembled both TX and RX, but I focused on TX code. From RX I got only telemetry ids. I will try to answer your questions but in next week - right now I am traveling my home town - so I will be out of access to my toys. From TX I have learned that the code is written horribly - most of time dependent actions are controlled from one place where some delays are added in between. So there is probability that they are not used at once. I will let you know.

**Comment by AleksanderAzizi** [comment]
@unitware Cool. Will it be added to the next release of BF?

**Comment by unitware** [comment]
@AleksanderAzizi no Idea, I'm only working with CF at the moment.

**Comment by marcroe** [comment]
@jakubk That sounds great. It would be very beneficial if you got some insight into the RX code, too. Can you send us the link for the TX firmware you are talking about? There are quite a few posts with different (modified) firmware versions on RCgroups and I don't know which one I am supposed to flash. 

@unitware @AleksanderAzizi  There's some testing going on. I basically just copied the Cleanflight code into BF code base and performed a test hover :)

**Comment by jakubk** [comment]
@marcroe 
Here it is:
https://www.rcgroups.com/forums/showpost.php?p=36186374&postcount=1143
If you want handle sensor data formatting just edit attached asm file. The method at 0xFB6C as parameters it takes:
R0 bufferPTR
R1 sensorID
R2 sensorValue
sprintf method is located at 0x19F8 - I have defined it in my asm file so you can use it. 
As reference you can use my SNRSensor or signalSensor they are very very simple one.
Default format method produces "%u" based string.
Currently sensor value is 2 bytes value, but I can change method getting telemetry value to merge 2 telemetry packets so we can get 4 bytes - going beyond that will require changes in RX and TX telemetry handling.  Telemetry names can be changed as well - just edit my bin file. If you change anything in binary check sum (last 2 bytes) must be recalculated - I think unofficial flash app can do it - but try to do it on your own. It is CRC-CCITT from beginning of the file until file length - 2.

**Comment by goldjacket** [comment]
I want to do this but I cant find a updater tool that will update the TX. They are all out there as updaters that are .exe that automatically update the firmware to whatever updater version you downloaded. There is an original updater but I cant seem to find it anywhere on the net. I found this Mod i6 alt_recovery updater v1_0.exe but it also just updates a internal firmware to the exe.  Is there another way to flash this  bin file?

**Comment by jakubk** [comment]
@goldjacket Updater can be found here https://github.com/ThomHPL/BaseI6_updater
Already packed version can be found here https://github.com/benb0jangles/FlySky-i6-Mod-/tree/master/10ch%20Timer%20Mod%20i6%20Updater/10ch_Timer_MOD_i6_Programmer_V1_3

**Comment by jakubk** [comment]
@unitware As promiced I have looked at disassembly of RX. Some time ago I have found IDA project created by stinkydiver so all credits to him.
There is only one subroutine containing references to IBUS communication (timer interrupt handler) - somewhere in this routine 
`LDR     R4, =byte_2000011C
LDRB    R0, [R4,#(byte_2000012B - 0x2000011C)]
CMP     R0, #0
BEQ     loc_800484A
BL      ibussensor
B       loc_8004A6C
loc_800484A:
BL      ibusservo
B       loc_8004A6C`

ibussensor and ibusservo are calling same method stinkydiver named this method ibusio. It is attached to this post. 
[ibus_io.zip](https://github.com/cleanflight/cleanflight/files/604660/ibus_io.zip)
I that this operation is atomic because USART receiver is being disabled as well as DMA transfers - when this routine is being executed. So I think your approach is correct. 
I think the key to recognize if it is servos or sensors time is checking when byte 2000012B is set to 0 and when it is set to 1. 
If you need full disassembly just let me know.

**Comment by unitware** [comment]
Thanks, I'll look deeper in this then I'm home again (travelling at the moment) but it looks promising. 

Is there any chance to build a version that copies(/or sends) all servo output to the sensor port at the same time? That would save some extra hardware.

Again, thanks!

Skickat från min iPhone

**Comment by jakubk** [comment]
@unitware 
It should be possible to use same line for sending data out of RX. The problem I can imagine is blocking the bus with incoming sensor data when servos packet must be send out but it depends on number of sensors. But I have to dig into asm code deeper. I have ordered Cleanflight compatible FC so soon I can start experimenting.
Have you ever thought about MAVlink telemetry format together with FlySky devices? Measurement packet format is similar and whole telemetry protocol will become much simpler - but much more data muse be transferred - I am not sure if wireless protocol can handle it. As first step anyhow I would like to have working simulated sensors.

Lately I have found bug in my assembly code for custom sensor handling - but it seems it was not used by anyone.

**Comment by povlhp** [comment]
If you look at the data with a traffic analyzer, you can see that the
sensor data is only sent in response to a request from the Rx. The 32-bytes
channel data is sent every 7ms or so, but takes only around half the time
to send.
I am not sure if we can connect the transmit line of the sensor iBus port
to the channel data iBus port so get the request packet. Otherwise it is
just using timing after channel data is received. Then we know we have a
window of around 3 ms to send data.

**Comment by jakubk** [comment]
@povlhp That was exactly what I meant  - but I was not aware of timing. Your calculations seems to be correct - transferring of 32 bytes is taking about 3456 uS (in standard serial transmission with 1 start bit, 8 bits of data, 1 bit parity, 2 bits stop at 115,200 bit/s). So remaining time can be used to request sensor data request is 4 bytes long so we can request up to 8 sensors. And one line is then fully utilized.  
I see 2 possibilities:
-because 2 way communication is performed only on sensors port we can use one serial line of RX for transmitting data and another one for receiving - so we can allow sending channel data (3.5 ms) and sensors requests (must be limited to 8 requests) on same line. We can send responses from FC to another one - but still one UART of FC will be used but both ports will be utilized. 
-using single line for channels data and 2 way sensors communication - it will allow us to handle in best case 2 sensors (4 bytes request and 6 response plus delay between request and response) - does not sound good for me.

**Comment by povlhp** [comment]
Do you have a logic analyzer ? If not, I might be able to get you a log of
communications at the logic level for both ports in parallel. I have
external battery sensor + RPM sensor (not used yet) I could hook up to an
iA6B.
The Salae Logic clones are cheap from China. Aroud $6-7 (24MHz, 8 channel)
You can use that to deduce timing.

**Comment by unitware** [comment]
I would love to see a log of all signals including the ones between and after the external sensors to verify the assumptions I made in the ibis telemetry implementation. 



Skickat från min iPhone

> 22 nov. 2016 kl. 12:00 skrev Povl H. Pedersen <notifications@github.com>:
> 
> Do you have a logic analyzer ? If not, I might be able to get you a log of
> communications at the logic level for both ports in parallel. I have
> external battery sensor + RPM sensor (not used yet) I could hook up to an
> iA6B.
> The Salae Logic clones are cheap from China. Aroud $6-7 (24MHz, 8 channel)
> You can use that to deduce timing.
>

**Comment by povlhp** [comment]
You can find a data dir now as part of my iBus2PPM project here:
https://github.com/povlhp/iBus2PPM

Raw data files from Sale Logic analyser, screenshot, and a text file
showing the bytes on the telemetry iBus wire with both the RPM + voltage
sensor. You can download the software for free at salae logic, and open the
raw files. (Mac, Windows, Linux). But some of the data you can find in txt
files.

To me it looks like the iA6B is actually pulling only one sensor between
each channel data packet out. So only 1 telemetry value should eb sent to
the iA6B for every channel data packet received.

Period between the packets is 7.7ms (both between iBus channel packets, and
between sensor request packets etc).
Period from end of data packets to first bit of request for sensor 0.844 ms.
From first byte of channel data to first byte of sensor answer: 1.85ms

When I look at data between the 2 iBus sensors, I see the requests from the
iA6B to both sensors (0xA1 & 0xA2), but I only see the 0xA2 answers. Not
the answer from the sensor in front. So it is not just using the same data
wire. But there is no measurable delay. The 2nd sensor sees the data at the
same time as the first, at least down to the same 0.844ms (with my 1MHz
sample rate)

Also logged during startup (between the sensors). Here is the data, seems
like some 0x0482... sequences as probes. Then 0x0402.... probably part of
sensor discovery, since it pulls them one at a time later on. I swapped
sensor order between the 2 logs of startup

Some of the startup sequence does not hit 2nd sensor.

I assume everything 0x04.. is from rx, 0x06 is from sensors. Can see the
discovery of the sensors as 0481, 0482, 0483. If the sensor is there, I see
the discovery packet twice. Maybe echoed by the receiving sensor to say it
is there. This is not part of the old Dave Borthwick PIC code.

Hope the data is useful for you and/or others.

**Comment by jakubk** [comment]
@povlhp thanks for your effort and logging all this data - I have logic analyser but I don't have any external sensors. I think your sensor's communication log confirms my deduction from disassembly about alternate usage of serial ports - I was not aware that only single sensor is requested - I have not check what exactly is send in ibus method because have just disassembly without RAM dump. Many thanks one more time. 
Echoing discovery message was already described by Dave Borthwick:
"The first discovered module echos back the message."
Dave has also provided some delay between sensor measurement request and response  -  712us
Lately I have found altitude sensor id used in i6X is 249 I have to check how exactly it is handled in RX because described protocol is not fitting at all this 2 byte addressing (also sensors 0x7c - Odometer 1, 0x7d - Odometer 2, 0x7e - Speed seems to have 2 bytes ID). Possibly IDs can be remapped internally. But it should be easy to find.

**Comment by povlhp** [comment]
The sensors are not really that expensive ($10 or so). Never used the RPM
one, just bought it (like too much other stuff) for future projects, mostly
for an easy way to measure static RPM on the table. Easy to build your own,
but time is a sparse resource.

The old PIC code from Dave I have does not have the first discovery
messages defined.

One thing that would be intersting is to know if we can just connect the 2
iBus data pins to each other, and thus get down to using only one serial
port in bidirectional mode on the FC. I did not get around to it, but will
see if I can get time to try log that. See if the signals from one wire
will interfere with the other.

I think it is worth checking if you can send any response back after the 04
request for data, or if it need to be from the right sensor. If this is the
case, we need the output from the sensor iBus. If not, just run
round-robin, sending one sensor after another in the timeslots. You might
be able to put 2 in between the channel packets, but there is also
considerations of CPU time and buffer size.

In iBus2PPM I played a lot with the PPM timing and other stuff, and
different FC implementations handle things different. The
CleanFlight/Betaflight was happy with dynamic frame length. I.e. don't wait
22.5ms between start of frames. Just send 2nd frame as soon as possible.
APM did not like that.

It would help some on latency, as you would decrease frame size from 22.5ms
to saving in average 0.5ms * 8 channels = 4ms.

In theory PPM could have even lower latency if I could send only channel
1..4 one every 2nd packet. Then the 2.5-3ms packet spacer, and next packet.
This would help even more. Or even only send channels 5..8 if any change.
But all this needs smarter receiving code in the FC. That would save 4*2ms
= 8 ms more. Bringing latency down from 22.5 to 12.5ms in average. So even
in PPM there is room for optimization, if the receiver code is updated.
This is a project I would consider at some point if I have time. But for
now iBus is way better.

**Comment by jakubk** [comment]
@povlhp 
We can not send multiple responses for request in stack firmware. There is check if response header is matching request header. The targetArrayPtr +48 is addess where response packet starts.
Here you can see fragment from rx code:
`   case 3u:
        if ( !(asSensors | (unsigned __int8)zeroIfSensors) )
          goto LABEL_33;
        sensorAddress = *(_BYTE *)(targetArrayPtr + 9);
        if ( ((sensorAddress << 8) | (*(_BYTE *)(sensorAddress + plus27) + 4) | val8000ifSensors | 0x2000) == *(_WORD *)(targetArrayPtr + 348) )
        {
          ((void (__fastcall *)(int, int))loc_8000230C)(16 * sensorAddress + targetArrayPtr + 44, targetArrayPtr + 350);
          v13 = (unsigned __int8)(*(_BYTE *)(targetArrayPtr + 9) + 1);
          *(_BYTE *)(targetArrayPtr + 9) = v13;
          v14 = *(_BYTE *)(targetArrayPtr + 8);
          if ( v13 > v14 )
            *(_BYTE *)(targetArrayPtr + 9) = 1;
          if ( v14 < 0xF && (unsigned int)(0x384B - *(_DWORD *)(targetArrayPtr + 304)) >= 0x1F4 )
            *(_BYTE *)targetArrayPtr = 1;
LABEL_33:
          *(_BYTE *)(targetArrayPtr + 10) = 0;
          break;`

**Comment by jakubk** [comment]
Some good news - RX is able to send to TX variable length telemetry reading - there is no 2 byte limitation!
The packet structure is following:
0xAC(byte)  SENSOR_ID (byte) SENSOR_INDEX (byte) PAYLOAD_LENGTH (byte) PAYLOAD (n bytes) ........... 
frame may contains number of sensors but total length can not exceed 28 bytes.
Handling in FS-i6 is missing.

**Comment by unitware** [comment]
Hi, a slight off topic question so that I might get started (please point me to good link if you have it, I have failed to google the answer).

I have recently updated to the 10channel mod and it went well (https://github.com/benb0jangles/FlySky-i6-Mod-/tree/master/10ch%20Mod%20i6%20Updater/10ch_MOD_i6_Programmer_V1_5) but I get only 6 channels on ppm (expected 8) And I cannot find the "Update receiver" menu item in "Rx setup" as I have understood. Has it moved somewhere else? How can I update the ia6b receiver?

**Comment by jakubk** [comment]
@unitware just update RX to FW 1.6
https://www.rcgroups.com/forums/showthread.php?2677694-How-to-upgrade-Flysky-Turnigy-iA6B-RX-to-firmware-1-6-with-a-ST-Link

**Comment by unitware** [comment]
ohh, I see, so I have to do it with a debugger (st-link)? I thought it was supposed to work over the air using the fw-upgrade button on the receiver?

**Comment by jakubk** [comment]
@unitware - yes or any other arm flash tool. It is working but not with i6 it has 64k flash and RX fw is about 32kB.

**Comment by unitware** [comment]
cool

meanwhile, I'm working on combining serial rx and ibus telemetry on the same pin without any specific mods to the flysky equipment. I'll start a new issue for that discussion so that this thread can focus on adding those sensors.

**Comment by inode-** [comment]
@qba667 I read the topic but I'm not sure that I understood everything.

To get all these sensors you have to modify also the firmware of the remote or the firmware of the receiver? or both?

Naturally it's needed also a patched version of the cleanflight that send the data to the receiver.

It will be possible to have more 3 telemetry data in the home screen?

**Comment by qba667** [comment]
@inode- RX for me is ia6b and TX is I6. To transmit data longer then 2 bytes for now small change is necessary in RX - currently only GPS data is defined as 4 byte but it is not implemented in TX right now - so you don't have to change anything in RX. I need some time or maybe sensor to understand what AD frame is exactly and what I should answer from TX. Anyhow all  2byte sensors should be working so far with my changed TX firmware. Some names are to long - this all must be adjusted. The goal is to use other font and display more sensors on main screen. That was my test from month ago:
![img_20161114_203256](https://cloud.githubusercontent.com/assets/3341647/21519621/024aed3a-ccef-11e6-962b-6435a9b6afa4.jpg)
To handle more sensors on main screen (up to 6) I have to remove sensor's index used to identify sensor - i think just ID is enough it will be done soon.
Sum up:
- to handle all 2byte sensors use my FW for I6 unit - it is experimental one. 
- 4 bytes sensors are not implemented in I6 because of lacking knowledge about AD frame - it must be send back to RX - without this operation RX will never propagate sensor data to TX
- implementation in flight controller must me done - reference code https://github.com/qba667/MAVLinkToIbus

**Comment by inode-** [comment]
@qba667 if needed I can provide sniffed data (with Saleae Logic Analyzer) from real IBUS external sensors (Voltage and RPM)

**Comment by unitware** [comment]
Is it possible to get the sensor data ouput on the tx serial port? I mean to get it to the computer while flying?

**Comment by qba667** [comment]
@unitware That should be easy to accomplish - few people asked me about it as well. But the format will be still IBUS (for now). I will try to implement it in the radio in next week.
@inode nice dumps are already created by @povlhp but thanks for your involvement.

**Comment by darmach** [comment]
@qba667 Sorry to spam here, just wanted to ask - As I see you're developing this using FS-I6X, FS-IA6B and naze32. I'm just starting with drones, I wanted to buy I6X, but now I'm wondering about support of that one. Maybe it would be better to buy I6 ? Will this work on SP Racing F3, or the older, naze32 is a must?

**Comment by unitware** [comment]
@darmach I have not tested i-bus on the F3 controllers (yet, will do in the week to come) but there should be no problems.

I have no idea of the difference between I6 and I6X, sorry.

**Comment by qba667** [comment]
@darmach 
I have worked with I6 - that is unit I own. Difference is in main-board (different MCU I6X has better one). There are also differences in firmware, the firmware of i6X is kind of extended one comparing to I6 - it can be also modified but I will not do it - I have no unit to perform any tests. What kind of telemetry do you need?

**Comment by darmach** [comment]
Ah, that's a tricky question - all of it! ,) But seriously - signal strength, voltage would be a must. But having a GPS position, speed, climbing speed or altitude would be really nice.

**Comment by qba667** [comment]
I would not invest in I6, because of hardware limitations. For now even with modified firmware i6X seems to be better choice - 10 channels and basic telemetry out of the box for 50$.

**Comment by darmach** [comment]
@qba_dx - I'd like to ask you few questions about these if you could spare a while. I don't want to spam here, would you ping me on my mail? darmach[on]poczta .pl

**Comment by qba667** [comment]
I have published FW allowing configuration of 6 sensors on main screen. Sending telemetry data to serial port is in progress.

**Comment by goldjacket** [comment]
"Sending telemetry data to serial port is in progress"  Does this mean I will be able to finally see my transmitter err% on my goggles so I know I know my limits while fpv-ing. If so yeaaa!!:)

**Comment by McGiverGim** [comment]
I've a FS-I6S with a FS-IA10B, I've done the hardware change and installed this pull-request:

https://github.com/cleanflight/cleanflight/pull/2554

Now the iBus telemetry is shown using only one harware serial port.

Can I "merge" or make something with some branch to test something of your changes?

**Comment by qba667** [comment]
@unitware 
I have added feature you asked once -  8 channels ppm to radio firmware. Serial sensor output is also almost ready - but I need some time for testing.

**Comment by inode-** [comment]
@qba667 any news about the patched firmware to send also other telemetry data (altitude etc)? And the 4 bytes data? (like gps cordinates).

Thank you!

**Comment by qba667** [comment]
@inode- yes I can write code for AC sensor (size 4 bytes and more) handling in C - so huge step forward - I was starting with manual assembly editing. I am currently experimenting with ALT sensor - multiple telemetry pages on main screen and few other things.. check RC groups topic for more details.

**Comment by qba667** [comment]
@unitware, @povlhp  Here you have firmware with "sticks" data and "sensors" data on sensors port.
![ibus_lines_together](https://cloud.githubusercontent.com/assets/3341647/25071993/927a0744-22c4-11e7-9d86-514d2d4beeb8.png)
![ibus_lines_together2](https://cloud.githubusercontent.com/assets/3341647/25071994/929e9e2e-22c4-11e7-90ef-4e3d7b9860b2.png)
[Same_serial_ia6b.zip](https://github.com/cleanflight/cleanflight/files/924346/Same_serial_ia6b.zip)

**Comment by unitware** [comment]
Cool, this should then work with "combunined ibus aeroalrx and telemetry" enabled but with only a wire (no need for diode norbresistance.

**Comment by qba667** [comment]
@unitware I tried software combined serial ports. Cleanflight hangs short after startup in such configuration.

**Comment by unitware** [comment]
Thanks, I have been flying combined on betaflight (using a wired combination with diode and resistor) and it worked. I have however switched receiver and FC since then so I have no real possibility to dig into this again.

**Comment by moirazakhatri1** [comment]
You just purchased a brand new car and you thought it would be great for your wedding day. You and your soon-to-be spouse can get from the ceremony, down the steps and into your waiting brand new automobile to head to the reception or even to begin your honeymoon. [Denver Party Bus](https://www.partybusdenver.net/denver-party-bus.html)
